Create a blog post about the Backports application (#686)
quarkusio/quarkusio.github.io: _posts/2020-08-11-backports-application.adoc

---
layout: post
title: From zero to hero on backporting pull-requests
date: 2020-08-11
tags: user-story
author: gastaldi
---

> = From zero to hero on backporting pull-requests

Reviewer: I would remove the dash for all the `pull-requests` occurrences.

> Before we dig into the solution, let me give you a quick explanation about our release process

Reviewer: ```suggestion
Before we dig into the solution, let me give you a quick explanation about our release process.
```

== Release process

> === Minor versions

Reviewer: I would say major versions here. In Quarkus, 1.x is really a major given we push a ton of new features to it.

Author: I think calling it major versions would be confusing to users, since semantic versioning 
states that the format is `MAJOR.MINOR.PATCH`, unless we put a paragraph saying that we don't follow semantic versioning (well, if we really followed it since the beginning, we'd probably have been in Quarkus `3000.0.0.Final` given the amount of API changes we introduced :smiley:)

Our master branch is always ready for the next minor release. That means that any pull-requests merged in there are available in the next minor release. This process is usually very smooth and involves no backporting.

=== Patch releases or CRs

> Whenever a new patched version (or a CR) is about to be released in Quarkus, our Release engineer starts backporting commits from pull-requests that were merged in the master branch.

Reviewer: ```suggestion
Whenever a new patched version (or a CR) is about to be released in Quarkus, our Release engineer starts backporting commits from pull-requests that were merged in the master branch.
```

The first CR is fine. That's all the subsequent releases that needs backporting (`.0.Final` included)

> How do they know which pull-request to grab? We have a `triage/backport?` label that our team add to pull-requests for features that would be worth having it in the upcoming release.

Reviewer: ```suggestion
How do they know which pull-request to grab? We have a `triage/backport?` label that our team add to pull-requests for features that would be worth having it in the upcoming patch release.
```

== What the app does?

> The application basically queries a GitHub repository's merged pull requests and closed issues (using the GraphQL API) containing a certain label and changes their milestones to one that is selected in the UI (and removing the certain label afterwards)

Reviewer: I think it's a bit limiting.

We need to explain:
- why we needed it: the number of pull requests to backport is enormous and doing it via the GitHub UI is cumbersome and takes a lot of time (click PR, copy paste commit hashes to cherry-pick, remove label, affect milestone, affect milestone to all fixes issues, next) + some limitation of the UI (it's not possible to sort by merge date to avoid conflicts)
- it also simplifies the cherry picking

=== Choosing a milestone
image::https://raw.githubusercontent.com/quarkusio/quarkus-backports/master/documentation/screenshots/index.png[]

=== Choosing the pull-requests to be backported
image::https://raw.githubusercontent.com/quarkusio/quarkus-backports/master/documentation/screenshots/backports.png[]

== Extensions

Here you can find a summary of the extensions used in the backports application:

- ArC (CDI): Manages the services lifecycle
- <<microprofile-config>>: To externalize properties

> - JAX-RS: Exposes an endpoint the UI can consume. Also serves the UI using Qute's Type-safe templates (see <<qute>>).

Reviewer: ```suggestion
- RESTEasy: Exposes an endpoint the UI can consume. Also serves the UI using Qute's Type-safe templates (see <<qute>>).
```
The extension is called RESTEasy.

> - <<qute>>: Transforms templates into a valid GraphQL query payload

Reviewer: well, it's used for all templating, HTML included.

- <<rest-client>>: To invoke the https://docs.github.com/en/graphql[GitHub's GraphQL endpoint]
- Cache: For https://github.com/quarkusio/quarkus-backports/blob/291868a9480601dd13b71703e7f8f73552109ded/src/main/java/io/quarkus/backports/GitHubService.java#L84[caching open Milestones] from the GitHub's GraphQL endpoint
- Hibernate Validator: For validating input
- Jackson (JSON library): To parse the result from the GitHub GraphQL endpoint

== Live Coding

> The live coding is a killer feature in Quarkus and provided a quick feedback while classes and methods were changed during development. As a rule of thumb, always use `mvn quarkus:dev` while developing a Quarkus application.

Reviewer: ```suggestion
The live coding is a killer feature in Quarkus and provided a quick feedback while classes and methods were changed during development. As a rule of thumb, always use `./mvnw quarkus:dev` while developing a Quarkus application.
```

[#microprofile-config]

> == Microprofile Config

Reviewer: ```suggestion
== MicroProfile Config
```
(and everywhere else)

Configuring which GitHub repository to use (to test or even for non-Quarkus repositories) and the GitHub authentication token (plus a different backport label if necessary) should be easy to configure without changing any source code.
Quarkus uses Microprofile Config, so we externalized these properties.

> Quarkus also supports `.env` files, which we used while testing. This made local testing easier by not requiring us to change the `application.properties` directly or setting any environment or system property before running the application

Reviewer: ```suggestion
Quarkus also supports `.env` files, which we used while testing. This made local testing easier by not requiring us to change the `application.properties` directly or setting any environment or system property before running the application.
```

We used https://quarkus.io/guides/qute#type-safe-templates[Type-safe templates] to generate the UI and the GraphQL queries.
[#rest-client]
== Rest Client

GraphQL, in a nutshell, means POSTing some JSON data to an HTTP endpoint and parsing the result as a JSON document. Simply that. The Microprofile REST Client is a good option to perform this task, so we came up with this:

```java
import javax.ws.rs.HeaderParam;
import javax.ws.rs.POST;
import javax.ws.rs.core.HttpHeaders;

import io.vertx.core.json.JsonObject;
import org.eclipse.microprofile.rest.client.inject.RegisterRestClient;

@RegisterRestClient(baseUri = "https://api.github.com/graphql")
public interface GraphQLClient {

    @POST
    JsonObject graphql(@HeaderParam(HttpHeaders.AUTHORIZATION) String authentication, JsonObject query);
}
```

In the GitHubService we can now consume the GraphQLClient object:

```java
    @Inject
    @RestClient
    GraphQLClient graphQLClient;
```

You can see how the client is invoked https://github.com/quarkusio/quarkus-backports/blob/291868a9480601dd13b71703e7f8f73552109ded/src/main/java/io/quarkus/backports/GitHubService.java#L85-L101[here]

== Next steps

> Given the nature of dynamic queries, we decided to not use the SmallRye GraphQL extension, but that can changed when the extension supports that feature.

Reviewer: ```suggestion
Given the nature of dynamic queries, we decided to not use the SmallRye GraphQL extension, but that can be changed when the extension supports that feature.
```

---
layout: post
title: From zero to hero on backporting pull requests
date: 2020-08-11
tags: user-story
author: gastaldi
---
= From zero to hero on backporting pull requests

Before we dig into the solution, let me give you a quick explanation about our release process.

== Release process

=== Minor versions
Our master branch is always ready for the next minor release. That means that any pull requests merged in there are available in the next major release. This process is usually very smooth and involves no backporting.

=== Patch releases or CRs
Whenever a new patched version (or a second CR) is about to be released in Quarkus, our Release engineer starts backporting commits from pull requests that were merged in the master branch.
How do they know which pull request to grab? We have a `triage/backport?` label that our team add to pull requests for features that would be worth having it in the upcoming patch release.

> == What the app does?

Reviewer 2: All of a sudden, you start talking about an app, assuming that I (as a reader) know it exists and am interested in it.

How about saying something like: "Considering the amount of backports, this gets tedious and error prone when done manually. So we decided to automate that, and built an application. Of course with Quarkus!"

The application basically queries a GitHub repository's merged pull requests and closed issues (using the GraphQL API) containing a certain label and changes their milestones to one that is selected in the UI (and removing the certain label afterwards)

=== Why we needed it?
The number of pull requests to backport is enormous and doing it via the GitHub UI is cumbersome and takes a lot of time (click PR, copy/paste commit hashes to cherry-pick, remove label, affect milestone, affect milestone to all fixes issues, next) plus some limitation of the UI (it's not possible to sort by merge date to avoid conflicts, for example)

> It also simplifies the cherry picking

Reviewer 2: Missing periods at end of sentences? Or just incomplete sentences?

---
layout: post
title: From zero to hero on backporting pull requests
date: 2020-08-11
tags: user-story
author: gastaldi
---
= From zero to hero on backporting pull requests

The Quarkus project moves fast and when we prepare bugfix releases, we usually have several dozens of pull requests to backport.

> The number of pull requests to backport is enormous and (as it was aforementioned) doing it via the GitHub UI is inconvenient and takes a lot of time (click PR, copy/paste commit hashes to cherry-pick, remove label, assign milestone, assign milestone to all fixed issues, next) plus some limitation of the UI (it's not possible to sort by merge date to avoid conflicts, for example).

Reviewer 2: Perhaps remove "(as it was aforementioned)", because it's no longer true (and it's superfluous).

So we decided to automate this work, and built an application.
Of course, with Quarkus!

> Before we dig into the solution, let me give you a quick explanation about our release process.

Reviewer 2: "explanation about" -> "explanation of"?

Reviewer 2: And generally someone should probably review this from grammar perspective. I see some things that I believe are tiny mistakes, but I'm not a native speaker, so... :-)

---
layout: post
title: From zero to hero on backporting pull requests
date: 2020-08-11
tags: user-story
author: gastaldi
---
= From zero to hero on backporting pull requests

> The Quarkus project moves fast and when we prepare bugfix releases, we usually have several dozens of pull requests to backport. Using the GitHub UI to do that is feasible but cumbersome.

Reviewer 2: These 2 paragraphs don't follow well.

And we say at least 3 times that we have many pull requests to process, which is why it's tedious. We should probably pull that here into the 1st paragraph, and then adjust the headings so that the text flows nicely. Something like:

```
= From zero to hero on backporting pull requests

The Quarkus project moves fast and when we prepare bugfix releases, we usually have several dozens of pull requests to backport.
Using the GitHub UI to do that is feasible but cumbersome.
So we decided to automate this work, and built an application.
Of course, with Quarkus!

Before we dig into it, let me give you a quick explanation of our release process.

== Release process

...

[remove the last paragraph "Considering the amount of backports ..."]

== How do we automate that

...

== Our experience with Quarkus

Yes, this would be instead of "Extensions", and all other headings (Live Coding, MP Config, Qute, Next steps) would be `===` under "Our experience"

...

== Thanks

...
```

That should work better I think.

=== Major and Minor bumps
Our master branch is always ready for the next major or minor release.  This process is usually very smooth and involves no backporting.

=== Patch or Classifier bumps
Whenever a new patched version (or a second CR) is about to be released in Quarkus, our Release team starts backporting commits from pull requests that were merged in the master branch.
How do they know which pull request to grab? We have a `triage/backport?` label that our team add to pull requests for features that would be worth having it in the upcoming patch release.

Considering the amount of backports, this gets tedious and error prone when done manually. So we decided to automate that, and built an application. Of course, with Quarkus!

== What the application does?
The application basically queries a GitHub repository's merged pull requests and closed issues (using https://docs.github.com/en/graphql[the GraphQL API exposed by GitHub]) containing a certain label and changes their milestones to one that is selected in the UI (and removing the certain label afterwards)

=== Why we needed it?

> The number of pull requests to backport is enormous and (as it was aforementioned) doing it via the GitHub UI is inconvenient and takes a lot of time (click PR, copy/paste commit hashes to cherry-pick, remove label, affect milestone, affect milestone to all fixes issues, next) plus some limitation of the UI (it's not possible to sort by merge date to avoid conflicts, for example).

Reviewer 2: Can we just not use the work "affect" when we mean "assign"?

Author: That's @gsmet's "fault" :wink:

Reviewer 2: I know, but it's one thing in a casual conversation and another in a blog post :-)

> Applying the changes introduced by these pull-requests (aka. cherry-picking) is also simplified.

Reviewer 2: OK, so does it just simplify working with the issues, or does it automate the backporting process itself, too?

Author: Just the issues, the backporting process is done manually done. I've changed the sentence a bit
